client-data/tools/document/document.js

```diff
@@ -0,0 +1,103 @@
+(function documents() { //Code isolation
+
+
+    var xlinkNS = "http://www.w3.org/1999/xlink";
+    var imgCount = 1;
+
+    function assert_count() {
+        if (Tools.svg.querySelectorAll("image").length >= Tools.server_config.MAX_DOCUMENT_COUNT) {
+            alert("Too many documents exist already");
+            throw new Error("Too many documents exist already");
+        }
+    }
+
+    function onstart() {
+        var fileInput = document.createElement("input");
+        fileInput.type = "file";
+        fileInput.accept = "image/*";
+        fileInput.click();
+        fileInput.addEventListener("change", function () {
+            assert_count();
+
+            var reader = new FileReader();
+            reader.readAsDataURL(fileInput.files[0]);
+
+            reader.onload = function (e) {
+                // use canvas to compress image
+                var image = new Image();
+                image.src = e.target.result;
+                image.onload = function () {
+
+                    assert_count();
+
+                    var uid = Tools.generateUID("doc"); // doc for document
+
+                    var ctx, size;
+                    var scale = 1;
+
+                    do {
+                        // Todo give feedback of processing effort
+
+                        ctx = document.createElement("canvas").getContext("2d");
+                        ctx.canvas.width = image.width * scale;
+                        ctx.canvas.height = image.height * scale;
+                        ctx.drawImage(image, 0, 0, image.width * scale, image.height * scale);
+                        var dataURL = ctx.canvas.toDataURL("image/webp", 0.7);
```

Comment:
We should probably use jpeg by default here... We can make it configurable, though.

https://caniuse.com/#feat=webp

Reply:
jpeg does not support transparency, which would require us to check what kind of image was uploaded and if it made use of the alpha channel.
If it is a png with transparency we would not be able to compress it other than by shrinking the dimensions.

Reply:
I would suggest using a webp polyfill, e.g. https://github.com/chase-moskal/webp-hero

Reply:
Let's maybe start with JPEG, and then add support for webp, with a polyfill, in a second pull request ? webp-hero won't work as-is with data urls...